### PR TITLE
[SPARK-25845][SQL] Fix MatchError for calendar interval type in range frame left boundary

### PR DESCRIPTION
**Pull request by jiangxb1987**
## What changes were proposed in this pull request?

WindowSpecDefinition checks start < last, but CalendarIntervalType is not comparable, so it would throw the following exception at runtime:

```
 scala.MatchError: CalendarIntervalType (of class org.apache.spark.sql.types.CalendarIntervalType$)      at 
 org.apache.spark.sql.catalyst.util.TypeUtils$.getInterpretedOrdering(TypeUtils.scala:58) at 
 org.apache.spark.sql.catalyst.expressions.BinaryComparison.ordering$lzycompute(predicates.scala:592) at 
 org.apache.spark.sql.catalyst.expressions.BinaryComparison.ordering(predicates.scala:592) at 
 org.apache.spark.sql.catalyst.expressions.GreaterThan.nullSafeEval(predicates.scala:797) at org.apache.spark.sql.catalyst.expressions.BinaryExpression.eval(Expression.scala:496) at org.apache.spark.sql.catalyst.expressions.SpecifiedWindowFrame.isGreaterThan(windowExpressions.scala:245) at 
 org.apache.spark.sql.catalyst.expressions.SpecifiedWindowFrame.checkInputDataTypes(windowExpressions.scala:216) at 
 org.apache.spark.sql.catalyst.expressions.Expression.resolved$lzycompute(Expression.scala:171) at 
 org.apache.spark.sql.catalyst.expressions.Expression.resolved(Expression.scala:171) at 
 org.apache.spark.sql.catalyst.expressions.Expression$$anonfun$childrenResolved$1.apply(Expression.scala:183) at 
 org.apache.spark.sql.catalyst.expressions.Expression$$anonfun$childrenResolved$1.apply(Expression.scala:183) at 
 scala.collection.IndexedSeqOptimized$class.prefixLengthImpl(IndexedSeqOptimized.scala:38) at scala.collection.IndexedSeqOptimized$class.forall(IndexedSeqOptimized.scala:43) at scala.collection.mutable.ArrayBuffer.forall(ArrayBuffer.scala:48) at 
 org.apache.spark.sql.catalyst.expressions.Expression.childrenResolved(Expression.scala:183) at 
 org.apache.spark.sql.catalyst.expressions.WindowSpecDefinition.resolved$lzycompute(windowExpressions.scala:48) at 
 org.apache.spark.sql.catalyst.expressions.WindowSpecDefinition.resolved(windowExpressions.scala:48) at 
 org.apache.spark.sql.catalyst.expressions.Expression$$anonfun$childrenResolved$1.apply(Expression.scala:183) at 
 org.apache.spark.sql.catalyst.expressions.Expression$$anonfun$childrenResolved$1.apply(Expression.scala:183) at 
 scala.collection.LinearSeqOptimized$class.forall(LinearSeqOptimized.scala:83)    
```

We fix the issue by only perform the check on boundary expressions that are AtomicType.

## How was this patch tested?

Add new test case in `DataFrameWindowFramesSuite`

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameWindowFramesSuite.scala [hunk 1]
```diff
@@ -267,6 +267,25 @@ class DataFrameWindowFramesSuite extends QueryTest with SharedSQLContext {
     )
   }
 
+  test("range between should accept interval values as left boundary") {
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameWindowFramesSuite.scala]: this is using intervals for both, rather than just left.

**Comment by jiangxb1987** [comment]
Also cc @gatorsmile @cloud-fan @hvanhovell

**Review by hvanhovell — APPROVED** [review]
LGTM

**Comment by jiangxb1987** [comment]
Merging to master, I can open another PR against 2.4 if required in the future.

**Pull request closed without merging**